### PR TITLE
Check the filesystem for isReadOnly

### PR DESCRIPTION
**Pull request by Lemmmy**
As mentioned in #1214. The IMount implementation is default to avoid breaking any existing implementations, and I have provided implementations for FileMount (actually checks the real filesystem file for canWrite), FileSystemWrapperMount (delegates) and SubMount (delegates). Everything else uses the default behaviour of returning `true` - I am not sure if returning `false` would be more sensible and just letting the operation silently fail or return an IOException (depending on the situation and operating system). As far as I understand for all existing mounts this will always get hit by the `writableMount` check anyway.

I haven't written tests for this yet as I am not sure it is sensible or possible to test due to filesystem differences, but as far as I have tested everything is working as intended on Windows and Linux.

To recap, the use case for this is for `edit` etc. to prevent saving if the filesystem/directory/file is marked by the OS as read-only. It relies on [File.canWrite](https://docs.oracle.com/javase/7/docs/api/java/io/File.html#canWrite()).

### CODE REVIEW

File: projects/core-api/src/main/java/dan200/computercraft/api/filesystem/IMount.java [hunk 1]
```diff
@@ -41,6 +41,17 @@ public interface IMount {
...
+     * Returns whether a file with a given path is read-only or not.
+     *
+     * @param path A file path in normalised format, relative to the mount location. ie: "programs/myprograms".
+     * @return If the file exists and is read-only.
+     * @throws IOException If an error occurs when checking whether the file is read-only.
+     */
+    default boolean isReadOnly(String path) throws IOException {
+        return true;
+    }
+
```
Comment (SquidDev) [inline projects/core-api/src/main/java/dan200/computercraft/api/filesystem/IMount.java]: I think it probably makes more sense to move this to `IWritableMount` and then default to returning `false`.

File: projects/core/src/main/java/dan200/computercraft/core/filesystem/FileMount.java [hunk 2]
```diff
@@ -143,6 +143,14 @@ public boolean isDirectory(String path) {
         return file.exists() && file.isDirectory();
     }
 
+    @Override
+    public boolean isReadOnly(String path) throws IOException {
+        if (!created()) return path.isEmpty();
+
+        var file = getRealPath(path);
+        return file.exists() && !file.canWrite();
```
Comment (SquidDev) [inline projects/core/src/main/java/dan200/computercraft/core/filesystem/FileMount.java]: I'm not sure about this. Just to confirm my understanding, if we've not created the mount, then `/` is considered read-only? However, this should only be the case if the actual "computercraft/computers" directory is read only, which hopefully isn't the case!

I wonder how we should handle missing files in general here. Currently `fs.isReadOnly("read-only-dir/missing-file")` will return `false`, which isn't great. A couple of possibilities I can think of:

 - Keep the behaviour here: it's not a big issue after all!

 - `isReadOnly` walks up the tree until it finds a file which exists, then returns that. My big concern here is someone doing `fs.isReadOnly(("/p"):rep(10000))`, which could cause us to spend ages in filesystem code. 

   Hrmr, maybe you could do a binary search over the file path looking for one which exists :thinking:. Feel like that might be a bit too complicated.

**Review by SquidDev — CHANGES_REQUESTED** [review]
Thank you so much for looking at this!

I've left one question about missing files below: this actually may be more complicated than I thought :/.

File: projects/core/src/main/java/dan200/computercraft/core/filesystem/FileMount.java (hunk 2 above)
Comment (Lemmmy) [reply projects/core/src/main/java/dan200/computercraft/core/filesystem/FileMount.java]: Looks like that was a total oversight with the created check there. I've removed the check entirely and now I believe the behaviour is correct. I went with traversing the tree in the same way `makeDirectory` does. In practice, 10000 repetitions is near-instantaneous, and 100000 repetitions will throw "Too long without yielding". Since `makeDirectory` would also have the same issue, I don't think this is a major concern for now.

javaw_Y0HY8Bis14

File: projects/core-api/src/main/java/dan200/computercraft/api/filesystem/IWritableMount.java [hunk 3]
```diff
@@ -78,4 +78,15 @@ public interface IWritableMount extends IMount {
...
+
+    /**
+     * Returns whether a file with a given path is read-only or not.
+     *
+     * @param path A file path in normalised format, relative to the mount location. ie: "programs/myprograms".
+     * @return If the file exists and is read-only.
+     * @throws IOException If an error occurs when checking whether the file is read-only.
+     */
+    default boolean isReadOnly(String path) throws IOException {
+        return true;
```
Comment (SquidDev) [inline projects/core-api/src/main/java/dan200/computercraft/api/filesystem/IWritableMount.java]:
```suggestion
        return false;
```

I think?

File: projects/core/src/main/java/dan200/computercraft/core/filesystem/FileMount.java [hunk 4]
```diff
@@ -143,6 +143,21 @@ public boolean isDirectory(String path) {
...
+        if (file.exists()) {
+            return !file.canWrite();
+        }
+
+        var parent = file.getParentFile();
+        while (!parent.exists() && parent != rootPath) {
+            parent = parent.getParentFile();
+        }
+
+        return !parent.canWrite();
```
Comment (SquidDev) [inline projects/core/src/main/java/dan200/computercraft/core/filesystem/FileMount.java]: I wonder if it's better to write this something like:

```java
while (true) {
  if (file.exists()) return !file.canWrite();
  if (file.equals(rootPath)) return false;
  file = file.getParentFile();
}
```

I think this is the same, except it means `isReadOnly` will return `false` when the root doesn't exist?

Actually, I wonder if it's worth adding some tests for these? Happy to do so myself if you'd rather not!

**Review by SquidDev — CHANGES_REQUESTED** [review]

File: projects/core-api/src/main/java/dan200/computercraft/api/filesystem/IWritableMount.java (hunk 3 above)
Comment (Lemmmy) [reply projects/core-api/src/main/java/dan200/computercraft/api/filesystem/IWritableMount.java]: Whoops, sorry, completely missed that one!

File: projects/core/src/main/java/dan200/computercraft/core/filesystem/FileMount.java (hunk 4 above)
Comment (Lemmmy) [reply projects/core/src/main/java/dan200/computercraft/core/filesystem/FileMount.java]: image

Seems to work! Thanks, that's much more concise.

**Comment by Lemmmy** [comment]
Tests would definitely be a good idea but I'm not sure if there's a way to comprehensively and reliably do so in a way that would work cross-platform; Windows for example doesn't actually have attribute-based read-only directories, unless you have some kind of mock filesystem set up or made certain parts of the test only run on certain platforms. I could be wrong though

**Comment by SquidDev** [comment]
:roll_eyes: Maybe I should start paying money for an external CI service. Local tests are definitely not this unreliable!
